ngx_http_js_challenge.c

```diff
@@ -335,32 +341,61 @@ int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
 }
 
 static ngx_int_t ngx_http_js_challenge_handler(ngx_http_request_t *r) {
-
     ngx_http_js_challenge_loc_conf_t *conf = ngx_http_get_module_loc_conf(r, ngx_http_js_challenge_module);
 
     if (!conf->enabled) {
         return NGX_DECLINED;
     }
 
-    unsigned long bucket = r->start_sec - (r->start_sec % conf->bucket_duration);
-    ngx_str_t addr = r->connection->addr_text;
+    // Check for X-REAL-IP header and fallback to connection IP if not present
+    ngx_str_t addr = r->connection->addr_text; // Default IP
+    ngx_list_part_t *part = &r->headers_in.headers.part;
+    ngx_table_elt_t *header = part->elts;
+    for (ngx_uint_t i = 0; i < part->nelts; i++) {
+        if (ngx_strncasecmp(header[i].key.data, (u_char *)"X-REAL-IP", header[i].key.len) == 0 && header[i].value.len > 0) {
+            addr = header[i].value;  // Use X-REAL-IP if available
+            break;
+        }
+        if (i == part->nelts - 1 && part->next != NULL) {
+            part = part->next;
+            header = part->elts;
+            i = 0;
+        }
+    }
 
+    // Extract User-Agent header
```

Comment:
you can access user agent from r->headers_in.user_agent struct

Reply:
I've cleaned up that code now.